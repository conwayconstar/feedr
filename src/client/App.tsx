import React from 'react';
import './App.css';
import { MenuHookInterface } from 'feedr';
import Menu from './components/Menu';
import Filters from './components/Filters';
import useMenu from './hooks/useMenu';
import Item from './components/Item';
import Loading from './components/Loading';
import Error from './components/Error';

const App: React.FC = () => {
  const {
    loading,
    items,
    menu,
    toggleItem,
    search,
    setSearch,
    error,
    dietarySelection,
  } : MenuHookInterface = useMenu();

  if (error) return <Error error={error} />;

  if (loading) return <Loading />;

  return (
    <div className="wrapper">
      <Menu count={menu.length} dietarySelection={dietarySelection} />
      <div className="container menu-builder">
        <div className="row">
          <div className="col-4">
            <Filters search={search} setSearch={setSearch} />
            <ul className="item-picker">
              {items.map((item) => (
                <Item key={item.id} item={item} toggleItem={toggleItem} />
              ))}
            </ul>
          </div>
          <div className="col-8">
            <h2>Menu preview</h2>
            <ul className="menu-preview">
              {menu.map((item) => (
                <Item key={item.id} item={item} toggleItem={toggleItem} menu />
              ))}
            </ul>
          </div>
        </div>
      </div>
    </div>
  );
};

export default App;
