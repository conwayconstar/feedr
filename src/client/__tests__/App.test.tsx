import { ReactWrapper, mount } from 'enzyme';
import React from 'react';
import { DietarySelection } from 'feedr';
import items from '../../server/items';
import { updateWrapper } from './testHelpers';
import App from '../App';

// @ts-ignore
global.fetch = jest.fn(() => Promise.resolve({
  json: () => Promise.resolve({ items }),
}));

describe('App loads and menu works', () => {
  let wrapper: ReactWrapper;

  it('Homepage matches snapshot', async () => {
    wrapper = mount(<App />);
    expect(wrapper.text()).toEqual('Loading...');
    await updateWrapper(wrapper);
    expect(wrapper.render()).toMatchSnapshot();
  });

  describe('App page functionality', () => {
    beforeEach(async () => {
      wrapper = mount(<App />);
      await updateWrapper(wrapper);
    });

    it('Should add an item to the menu', () => {
      const item = wrapper.find('.item-picker Item').first();
      item.simulate('click');
      wrapper.update();
      const menuItems = wrapper.find('.menu-preview Item');
      expect(wrapper.find('.item-picker Item').length).toEqual(19);
      expect(menuItems.length).toEqual(1);
      expect(menuItems.first().prop('item')).toEqual({
        id: 1001,
        name: 'Kale Caesar Pasta, Turmeric Satay Broccoli & Lemon Cashew Greens',
        dietaries: ['v', 've', 'df', 'gf', 'n!'],
      });
    });

    it('Should remove an item', () => {
      const itemEls = wrapper.find('.item-picker Item');
      for (let i = 0; i < 4; i += 1) {
        itemEls.at(i).simulate('click');
        wrapper.update();
      }
      const menuItems = wrapper.find('.menu-preview Item');
      expect(wrapper.find('.item-picker Item').length).toEqual(16);
      expect(menuItems.length).toEqual(4);
      menuItems.at(1).find('button').simulate('click');
      wrapper.update();
      expect(wrapper.find('.menu-preview Item').length).toEqual(3);
      expect(wrapper.find('.item-picker Item').length).toEqual(17);
      expect(wrapper.find('.menu-preview').render()).toMatchSnapshot();
    });

    it('Should add up dietary info', () => {
      const itemEls = wrapper.find('.item-picker Item');
      for (let i = 0; i < 4; i += 1) {
        itemEls.at(i).simulate('click');
        wrapper.update();
      }

      const menu = wrapper.find('Menu');
      const dietarySelection: DietarySelection = menu.prop('dietarySelection');
      Object.keys(dietarySelection).forEach((diet, i) => {
        const dietary = menu.find('.menu-summary-right > span').at(i);
        expect(dietary.find('.dietary-count').text()).toEqual(`${dietarySelection[diet]}x`);
        expect(dietary.find('.dietary').text()).toEqual(diet);
      });
    });

    it('Should filter items which contain Cuban Beef in the name', async () => {
      const input = wrapper.find('input');
      input.simulate('change', { target: { value: 'Cuban Beef' } });
      wrapper.update();

      const itemEls = wrapper.find('.item-picker Item');
      expect(itemEls.length).toEqual(2);
      expect(itemEls.at(0).find('h2').text()).toEqual('Cuban Beef, Brown Rice & Quinoa, Green Pepper & Butterbean Salad');
      expect(itemEls.at(1).find('h2').text()).toEqual('Cuban Beef, Herby Potatoes & Turmeric Satay Broccoli');
    });
  });

  it('Should catch error', async () => {
    const apiError = new Error('The API is down :(');
    // @ts-ignore
    fetch.mockImplementationOnce(() => Promise.reject(apiError));
    wrapper = mount(<App />);
    await updateWrapper(wrapper);

    expect(wrapper.text()).toEqual(apiError.message);
  });
});
