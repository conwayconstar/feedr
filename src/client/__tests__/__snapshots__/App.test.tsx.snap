// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App loads and menu works App page functionality Should remove an item 1`] = `
<ul
  class="menu-preview"
>
  <li
    class="item"
  >
    <h2>
      Kale Caesar Pasta, Turmeric Satay Broccoli & Lemon Cashew Greens
    </h2>
    <p>
      <span
        class="dietary"
      >
        v
      </span>
      <span
        class="dietary"
      >
        ve
      </span>
      <span
        class="dietary"
      >
        df
      </span>
      <span
        class="dietary"
      >
        gf
      </span>
      <span
        class="dietary"
      >
        n!
      </span>
    </p>
    <button
      class="remove-item"
      type="button"
    >
      x
    </button>
  </li>
  <li
    class="item"
  >
    <h2>
      Dill & Swiss Chard Potato Cakes, Summer Tabbouleh & Roasted Roots
    </h2>
    <p>
      <span
        class="dietary"
      >
        gf
      </span>
      <span
        class="dietary"
      >
        df
      </span>
      <span
        class="dietary"
      >
        v
      </span>
      <span
        class="dietary"
      >
        ve
      </span>
      <span
        class="dietary"
      >
        n!
      </span>
    </p>
    <button
      class="remove-item"
      type="button"
    >
      x
    </button>
  </li>
  <li
    class="item"
  >
    <h2>
      Hake & Smoky Chickpeas, Herby Potatoes & Turmeric Satay Broccoli
    </h2>
    <p>
      <span
        class="dietary"
      >
        df
      </span>
      <span
        class="dietary"
      >
        gf
      </span>
      <span
        class="dietary"
      >
        rsf
      </span>
      <span
        class="dietary"
      >
        n!
      </span>
    </p>
    <button
      class="remove-item"
      type="button"
    >
      x
    </button>
  </li>
</ul>
`;

exports[`App loads and menu works Homepage matches snapshot 1`] = `
<div
  class="wrapper"
>
  <div
    class="menu-summary"
  >
    <div
      class="container"
    >
      <div
        class="row"
      >
        <div
          class="col-6 menu-summary-left"
        >
          <span
            class="count"
          >
            0 items
          </span>
        </div>
        <div
          class="col-6 menu-summary-right"
        />
      </div>
    </div>
  </div>
  <div
    class="container menu-builder"
  >
    <div
      class="row"
    >
      <div
        class="col-4"
      >
        <div
          class="filters"
        >
          <input
            class="form-control"
            placeholder="Name"
            value=""
          />
        </div>
        <ul
          class="item-picker"
        >
          <li
            class="item"
          >
            <h2>
              Kale Caesar Pasta, Turmeric Satay Broccoli & Lemon Cashew Greens
            </h2>
            <p>
              <span
                class="dietary"
              >
                v
              </span>
              <span
                class="dietary"
              >
                ve
              </span>
              <span
                class="dietary"
              >
                df
              </span>
              <span
                class="dietary"
              >
                gf
              </span>
              <span
                class="dietary"
              >
                n!
              </span>
            </p>
          </li>
          <li
            class="item"
          >
            <h2>
              Hake & Smoky Chickpeas, Brown Rice & Quinoa, Roasted Roots
            </h2>
            <p>
              <span
                class="dietary"
              >
                gf
              </span>
              <span
                class="dietary"
              >
                df
              </span>
              <span
                class="dietary"
              >
                rsf
              </span>
            </p>
          </li>
          <li
            class="item"
          >
            <h2>
              Dill & Swiss Chard Potato Cakes, Summer Tabbouleh & Roasted Roots
            </h2>
            <p>
              <span
                class="dietary"
              >
                gf
              </span>
              <span
                class="dietary"
              >
                df
              </span>
              <span
                class="dietary"
              >
                v
              </span>
              <span
                class="dietary"
              >
                ve
              </span>
              <span
                class="dietary"
              >
                n!
              </span>
            </p>
          </li>
          <li
            class="item"
          >
            <h2>
              Hake & Smoky Chickpeas, Herby Potatoes & Turmeric Satay Broccoli
            </h2>
            <p>
              <span
                class="dietary"
              >
                df
              </span>
              <span
                class="dietary"
              >
                gf
              </span>
              <span
                class="dietary"
              >
                rsf
              </span>
              <span
                class="dietary"
              >
                n!
              </span>
            </p>
          </li>
          <li
            class="item"
          >
            <h2>
              Cuban Beef, Brown Rice & Quinoa, Green Pepper & Butterbean Salad
            </h2>
            <p>
              <span
                class="dietary"
              >
                gf
              </span>
              <span
                class="dietary"
              >
                df
              </span>
            </p>
          </li>
          <li
            class="item"
          >
            <h2>
              Thai Chicken, Brown Rice & Quinoa, Lemon & Cashew Greens
            </h2>
            <p>
              <span
                class="dietary"
              >
                gf
              </span>
              <span
                class="dietary"
              >
                df
              </span>
              <span
                class="dietary"
              >
                rsf
              </span>
              <span
                class="dietary"
              >
                n!
              </span>
            </p>
          </li>
          <li
            class="item"
          >
            <h2>
              Cuban Beef, Herby Potatoes & Turmeric Satay Broccoli
            </h2>
            <p>
              <span
                class="dietary"
              >
                gf
              </span>
              <span
                class="dietary"
              >
                df
              </span>
              <span
                class="dietary"
              >
                n!
              </span>
            </p>
          </li>
          <li
            class="item"
          >
            <h2>
              Thai Chicken, Brown Rice & Quinoa & Turmeric Satay Broccolli
            </h2>
            <p>
              <span
                class="dietary"
              >
                gf
              </span>
              <span
                class="dietary"
              >
                df
              </span>
              <span
                class="dietary"
              >
                rsf
              </span>
              <span
                class="dietary"
              >
                n!
              </span>
            </p>
          </li>
          <li
            class="item"
          >
            <h2>
              Mackerel Salad Box
            </h2>
            <p>
              <span
                class="dietary"
              >
                gf
              </span>
              <span
                class="dietary"
              >
                df
              </span>
            </p>
          </li>
          <li
            class="item"
          >
            <h2>
              Pasta Salad Box
            </h2>
            <p>
              <span
                class="dietary"
              >
                ve
              </span>
              <span
                class="dietary"
              >
                v
              </span>
              <span
                class="dietary"
              >
                gf
              </span>
              <span
                class="dietary"
              >
                df
              </span>
            </p>
          </li>
          <li
            class="item"
          >
            <h2>
              Tuna Salad Box
            </h2>
            <p />
          </li>
          <li
            class="item"
          >
            <h2>
              Coconut Blondie with Dark Chocolate
            </h2>
            <p>
              <span
                class="dietary"
              >
                gf
              </span>
              <span
                class="dietary"
              >
                v
              </span>
            </p>
          </li>
          <li
            class="item"
          >
            <h2>
              Cinnamon Swirl Muffin
            </h2>
            <p>
              <span
                class="dietary"
              >
                gf
              </span>
              <span
                class="dietary"
              >
                df
              </span>
            </p>
          </li>
          <li
            class="item"
          >
            <h2>
              Apple, Date and Walnut Muffin
            </h2>
            <p>
              <span
                class="dietary"
              >
                df
              </span>
              <span
                class="dietary"
              >
                gf
              </span>
              <span
                class="dietary"
              >
                n!
              </span>
              <span
                class="dietary"
              >
                v
              </span>
            </p>
          </li>
          <li
            class="item"
          >
            <h2>
              Blueberry and Banana Muffin
            </h2>
            <p>
              <span
                class="dietary"
              >
                v
              </span>
              <span
                class="dietary"
              >
                df
              </span>
              <span
                class="dietary"
              >
                gf
              </span>
            </p>
          </li>
          <li
            class="item"
          >
            <h2>
              Ombar Chocolate Bar - Dark 72%
            </h2>
            <p>
              <span
                class="dietary"
              >
                df
              </span>
              <span
                class="dietary"
              >
                ve
              </span>
              <span
                class="dietary"
              >
                gf
              </span>
            </p>
          </li>
          <li
            class="item"
          >
            <h2>
              Ombar Bar - Milk
            </h2>
            <p>
              <span
                class="dietary"
              >
                ve
              </span>
              <span
                class="dietary"
              >
                gf
              </span>
            </p>
          </li>
          <li
            class="item"
          >
            <h2>
              Dark Chocolate Brownie
            </h2>
            <p>
              <span
                class="dietary"
              >
                v
              </span>
              <span
                class="dietary"
              >
                gf
              </span>
            </p>
          </li>
          <li
            class="item"
          >
            <h2>
              Mangajo Pomegranate
            </h2>
            <p>
              <span
                class="dietary"
              >
                ve
              </span>
              <span
                class="dietary"
              >
                df
              </span>
              <span
                class="dietary"
              >
                gf
              </span>
            </p>
          </li>
          <li
            class="item"
          >
            <h2>
              Mangajo Acai
            </h2>
            <p>
              <span
                class="dietary"
              >
                ve
              </span>
              <span
                class="dietary"
              >
                v
              </span>
              <span
                class="dietary"
              >
                gf
              </span>
            </p>
          </li>
        </ul>
      </div>
      <div
        class="col-8"
      >
        <h2>
          Menu preview
        </h2>
        <ul
          class="menu-preview"
        />
      </div>
    </div>
  </div>
</div>
`;
